### PR TITLE
done

### PR DESCRIPTION
**Pull request by jo-pol**
last line causes premature termination (exit) of test the process
so I suppose I did something wrong in for comprehension of reportWriting

### CODE REVIEW

File: src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala [hunk 1]
```diff
@@ -16,97 +40,70 @@ object FileIOSolution extends App {
...
   val reportWriting: ManagedResource[Unit] = for {
-    rs <- reports
-    (report1, report2) = rs
-    r1 <- Using.fileWriter()(report1File)
-    r2 <- Using.fileWriter()(report2File)
-  } yield {
-    writeReport(report1, r1)
-    writeReport(report2, r2)
-  }
+    reps <- reports
```
Comment (jo-pol) [inline src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala:99]: Debug hint: set breakpoints on all lines from 99-108, don't step over but resume

File: src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala [hunk 2]
```diff
@@ -16,97 +40,70 @@ object FileIOSolution extends App {
...
-    read(in) {
-      case Array(id, first, last) => Customer(id, first, last)
-    }
+    def create(l: Array[String]) = Customer(l(0), l(1), l(2))
+    read(in, create)
   }
 
   def readProducts(in: InputStream): List[Product] = {
-    read(in) {
-      case Array(id, title, price) => Product(id, title, price.toDouble)
```
Comment (jo-pol) [inline src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala:38]: much better

File: src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala [hunk 3]
```diff
@@ -16,97 +40,70 @@ object FileIOSolution extends App {
...
   }
 
-  def report1(orders: List[Order], products: List[Product], customers: List[Customer]): List[String] = {
-    orders.map(order => {
-      val reportLine = for {
-        product <- products.find(product => product.id == order.productId)
-        customer <- customers.find(customer => customer.id == order.customerId)
-      } yield s"${customer.firstName} ${customer.lastName} wants ${order.amount}x ${product.title}"
-
-      reportLine.getOrElse(s"CORRUPT ORDER: $order")
```
Comment (jo-pol) [inline src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala:55]: you didn't want us to do error handling

File: src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala [hunk 4]
```diff
@@ -16,97 +40,70 @@ object FileIOSolution extends App {
   val report1File = new File("report1.txt")
   val report2File = new File("report2.txt")
 
-  // as all three files need to be read in the same way, we can give a general function `read`
-  // with a transformer function of type `Array[String] => T`
-  def read[T](in: InputStream)(transform: Array[String] => T): List[T] = {
-    Source.fromInputStream(in)
```
Comment (jo-pol) [inline src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala:22]: Has a close so should also be wrapped in scala-arm.

File: src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala [hunk 5]
```diff
@@ -16,97 +40,70 @@ object FileIOSolution extends App {
...
-    pIn <- Using.fileInputStream(productFile)
-    oIn <- Using.fileInputStream(orderFile)
-  } yield generateReports(cIn, pIn, oIn)
+    orders <- Using.fileInputStream(orderFile) //.acquireAndGet(FileIO.readOrders)
+    customers <- Using.fileInputStream(customerFile) //.acquireAndGet(FileIO.readCustomers)
+    products <- Using.fileInputStream(productFile) //.acquireAndGet(FileIO.readProducts)
+  } yield generateReports(customers, products, orders)
 
+  import resource._
   val reportWriting: ManagedResource[Unit] = for {
```
Comment (rvanheest) [inline src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala:98]: was just playing with this and got something different, but maybe nicer: just thought of using a 'foreach' loop here. It appears that if you do not write `yield`, you get the behavior from `foreach`. With this you don't need this weird `reportWriting: ManagedResource[Unit]`, nor do you need the `acquireAndGet`. Just use the following:

``` scala
for {
  rs <- reports
  (report1, report2) = rs
  r1 <- Using.fileWriter()(report1File)
  r2 <- Using.fileWriter()(report2File)
} {
  writeReport(report1, r1)
  writeReport(report2, r2)
  println("done")
}
```

As I understand it right now, the lines within the first block are kind of the same and with the 'for-yield': a nicer translation of using `map` and `flatMap`, whereas the second block represents a `foreach` statement...

File: src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala [hunk 6]
```diff
@@ -16,97 +40,70 @@ object FileIOSolution extends App {
...
-    r1 <- Using.fileWriter()(report1File)
-    r2 <- Using.fileWriter()(report2File)
-  } yield {
-    writeReport(report1, r1)
-    writeReport(report2, r2)
-  }
+    reps <- reports
+    writer1 <- managed(new FileWriter(new File("target/report1.txt")))
+    writer2 <- managed(new FileWriter(new File("target/report2.txt")))
+    bufferedWriter1 <- managed(new BufferedWriter(writer1))
```
Comment (rvanheest) [inline src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala:102]: I see why you would do both the `FileWriter` and the `BufferedWriter`, but isn't this just what `fileWriter` is already doing?

File: src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala [hunk 7]
```diff
@@ -16,97 +40,70 @@ object FileIOSolution extends App {
...
+  } yield generateReports(customers, products, orders)
 
+  import resource._
   val reportWriting: ManagedResource[Unit] = for {
-    rs <- reports
-    (report1, report2) = rs
-    r1 <- Using.fileWriter()(report1File)
-    r2 <- Using.fileWriter()(report2File)
-  } yield {
-    writeReport(report1, r1)
```
Comment (jo-pol) [inline src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala:107]: From the 'almost done' revision I understood a `FileWriter` was not accepted as a `BufferedWriter`. The method `Using.fileWriter` is a nice short cut, but the name is misleading.

File: src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala [hunk 8]
```diff
@@ -16,97 +40,70 @@ object FileIOSolution extends App {
   val report1File = new File("report1.txt")
   val report2File = new File("report2.txt")
```
Comment (jo-pol) [inline src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala:41]: Write to `target/report1.txt` so you won't need a git-ignore.

File: src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala [hunk 9]
```diff
@@ -16,97 +40,70 @@ object FileIOSolution extends App {
...
 
-  // as all three files need to be read in the same way, we can give a general function `read`
-  // with a transformer function of type `Array[String] => T`
-  def read[T](in: InputStream)(transform: Array[String] => T): List[T] = {
```
Comment (jo-pol) [inline src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala:21]: Why currying?

File: src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala (hunk 5 above)
Comment (jo-pol) [reply src/main/scala/workshop3/assignments/solutions/FileIOSolution.scala:98]: I suppose it depends on when you want what to be executed. See the debugging hint.

**Pull request merged**